### PR TITLE
Fix #11193

### LINKED ISSUE #11193: ArgumentException from Microsoft.CodeAnalysis.Completion.Providers.SymbolCompletionItem.Create
**Version Used**: 1.3.0.60508

**Steps to Reproduce**:

1. Git clone roslyn-analyzers git repo and checkout the sources from 8th May
2. cibuild - to build the repo
3. Open src\Analyzers.sln
4. Open file: \src\Microsoft.QualityGuidelines.Analyzers\Core\UseLiteralsWhereAppropriate.cs
5. Go to line 73 and delete all text after 'fieldInitializerValue.Type' on that line in the if condition, including the && operator.
6. Enter '.SpecialType ==' after 'fieldInitializerValue.Type' and hit space - you should hit this crash

**Expected Behavior**:

No Crash

**Actual Behavior**:

VS Crashes with `ArgumentException`

```
   SyntaxTree '...' not found to remove

   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CommonGetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility)
   at Microsoft.CodeAnalysis.Compilation.GetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility)
   at Microsoft.CodeAnalysis.SymbolId.GetInteriorSymbols(ISymbol containingSymbol, List`1 symbols)
   at Microsoft.CodeAnalysis.SymbolId.GetInteriorSymbolOccurrence(ISymbol symbol)
   at Microsoft.CodeAnalysis.SymbolId.Generator.EncodeOccurrence(ISymbol symbol)
   at Microsoft.CodeAnalysis.SymbolId.Generator.VisitLocal(ILocalSymbol symbol)
   at Microsoft.CodeAnalysis.SymbolId.Generator.Visit(ISymbol symbol)
   at Microsoft.CodeAnalysis.SymbolId.CreateId(ISymbol symbol)
   at Microsoft.CodeAnalysis.Completion.Providers.SymbolCompletionItem.EncodeSymbol(ISymbol symbol)
   at Microsoft.CodeAnalysis.Completion.Providers.SymbolCompletionItem.EncodeSymbols(IReadOnlyList`1 symbols)
   at Microsoft.CodeAnalysis.Completion.Providers.SymbolCompletionItem.Create(String displayText, TextSpan span, IReadOnlyList`1 symbols, Int32 contextPosition, Int32 descriptionPosition, String sortText, String insertionText, Nullable`1 glyph, String filterText, Boolean preselect, SupportedPlatformData supportedPlatforms, Boolean isArgumentName, ImmutableDictionary`2 properties, ImmutableArray`1 tags, CompletionItemRules rules)
   at Microsoft.CodeAnalysis.Completion.Providers.AbstractSymbolCompletionProvider.CreateItem(String displayText, String insertionText, Int32 position, List`1 symbols, AbstractSyntaxContext context, TextSpan span, Boolean preselect, SupportedPlatformData supportedPlatformData)
   at Microsoft.CodeAnalysis.Completion.Providers.AbstractSymbolCompletionProvider.CreateItem(String displayText, String insertionText, Int32 position, List`1 symbols, AbstractSyntaxContext context, TextSpan span, Dictionary`2 invalidProjectMap, List`1 totalProjects, Boolean preselect)
   at Microsoft.CodeAnalysis.Completion.Providers.AbstractSymbolCompletionProvider.<>c__DisplayClass6_0.<CreateItems>b__3(IGrouping`2 g)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.CodeAnalysis.Completion.Providers.AbstractSymbolCompletionProvider.CreateItems(Int32 position, IEnumerable`1 symbols, TextSpan span, AbstractSyntaxContext context, Dictionary`2 invalidProjectMap, List`1 totalProjects, Boolean preselect)
   at Microsoft.CodeAnalysis.Completion.Providers.AbstractSymbolCompletionProvider.<GetItemsWorkerAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Completion.Providers.AbstractSymbolCompletionProvider.<ProvideCompletionsAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Completion.CompletionServiceWithProviders.<GetProviderCompletionsAsync>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Completion.CompletionServiceWithProviders.<GetCompletionsAsync>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.SpellCheck.AbstractSpellCheckCodeFixProvider`1.<CreateSpellCheckCodeIssueAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.SpellCheck.AbstractSpellCheckCodeFixProvider`1.<RegisterCodeFixesAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.<PerformActionAsync>d__2.MoveNext()
```

### PR DESCRIPTION
**Pull request by rchande**
When a SymbolId is created from a local symbol that comes from a
SpeculativeSemanticModel, GetInteriorSymbols crashes because it assumes
taht the LocalSymbol and its containing MethodSymbol have the same
SyntaxTree. With the change to use SymbolId in Completion, this causes
frequent crashes. Augment GetInteriorSymbols to not crash; this may cause
the completion tooltip to show incorrect type data if a local with the
same name is created multiple times in a method.

### CODE REVIEW

**Comment by rchande** [comment]
Tagging @dotnet/roslyn-ide for review.

File: src/Workspaces/Core/Portable/SymbolId/SymbolId.cs [hunk 1]
```diff
@@ -1027,7 +1037,7 @@ private void ParseArrayAndPointerTypes(List<ISymbol> symbols)
 
             private void ConstructArray(List<ISymbol> symbols, int bounds)
             {
-                for (int i = 0; i < symbols.Count; )
+                for (int i = 0; i < symbols.Count;)
```
Comment (rchande) [inline src/Workspaces/Core/Portable/SymbolId/SymbolId.cs]: Formatting change.

File: src/Workspaces/CoreTest/SymbolIdTests.cs [hunk 2]
```diff
@@ -427,6 +427,54 @@ public class C<S, T>
...
+}";
+            int position;
+            string text;
+            MarkupTestFile.GetPosition(markup, out text, out position);
+
+            var sourceText = SourceText.From(text);
+            var workspace = new AdhocWorkspace();
+            var project = workspace.AddProject("Test", LanguageNames.CSharp);
+            var document = workspace.AddDocument(project.Id, "testdocument", sourceText);
+            //var document = workspace.CurrentSolution.GetDocument(id);
```
Comment (jasonmalinowski) [inline src/Workspaces/CoreTest/SymbolIdTests.cs]: Looks dead to me.

File: src/Workspaces/CoreTest/SymbolIdTests.cs [hunk 3]
```diff
@@ -427,6 +427,54 @@ public class C<S, T>
...
+            var basemethod2 = tree2.FindTokenOnLeftOfPosition(position, CancellationToken.None).GetAncestor<BaseMethodDeclarationSyntax>();
+
+            var service = new CSharpSemanticFactsService();
+            SemanticModel testModel;
+            var m = service.TryGetSpeculativeSemanticModel(firstModel, basemethod1, basemethod2, out testModel);
+
+            var xSymbol = testModel.LookupSymbols(position).First(s => s.Name == "x");
+
+            // This should not throw an exception.
+            SymbolId.CreateId(xSymbol);
```
Comment (jasonmalinowski) [inline src/Workspaces/CoreTest/SymbolIdTests.cs]: Can you assert what you expect to be getting out of this now? null/non-null at least?

**Comment by jasonmalinowski** [comment]
It seems strange now that SymbolId works inconsistently if it's handed a speculative vs. normal semantic model. It also seems strange that the local symbols from a speculative model say they're from a compilation that they're really not. Should the fix be either on the compiler side for this, or should completion just not be asking for SymbolIds when there's a local? (what happens now that it won't get them?)

**Comment by rchande** [comment]
@jasonmalinowski: Yes, the real fix is for the compiler to enable getting the locals declared in a method , which is tracked by https://github.com/dotnet/roslyn/issues/11349. This change is to unblock dogfooding.

If GetInteriorSymbols doesn't provide any symbols, completion could potentiallly treat locals in a method as references to the first declaration of a local with that name, in which case the tooltip might show the wrong type. However, I haven't actually been able to make that happen.

**Comment by CyrusNajmabadi** [comment]
I'm torn about this change.   I don't like inferences in code which may mask real bugs.  For example, the PR says:

> When a SymbolId is created from a local symbol that comes from a
SpeculativeSemanticModel, GetInteriorSymbols crashes

However, what the code does is just check SyntaxTrees.  This means it might mask a bug where we're *not* using a SpeculativeSemanticModel, but we end up with mismatched trees for some reason.

Is it possible for an Symbol to expose if it was created from a Speculative SemanticModel or not?  Could we then just skip these symbols in that case?  Or would that be too heavyweight?  Is it more that we actually just want to ignore these specific locations, while still working on the other locations?

**Comment by rchande** [comment]
@CyrusNajmabadi I don't think we have a good way to make Symbols say what kind of SemanticModel they're from, though that might be nice. The problem is that this piece of code needs to go from the Symbol to the Compilation to figure out what locals are declared. There's no direct pointer from a Symbol to the Compilation it came from so this code has to start with the Compilation for the ContainingAssembly, which wasn't rebuilt by the speculative model to match the new set of trees.

For now I would probably argue that we should take this to fix the crashes and get rid of this code entirely when we add the compiler API to get the interior symbols. A (very labor intensive) alternative might be to plumb a bit all the way through the SymbolID visitor to remember whether the model was speculative.

**Comment by jasonmalinowski** [comment]
How big is the compiler fix? Why can't we just do the right thing instead?

**Comment by jmarolf** [comment]
@CyrusNajmabadi an alternate fix for this would be to have the symbol be able to indicate that it came for a semantic model and allow you to retrieve the speculative semantic model that created it, it similar to how [ISourceAssemblySymbol](http://source.roslyn.io/#Microsoft.CodeAnalysis/Symbols/ISourceAssemblySymbol.cs) allows you to get the compilation.

@jasonmalinowski talking with @gafter there are some design concerns that need to be worked out to provide this API.  Its unclear if this is something we can get for update 3.

**Comment by jmarolf** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
If we take this approach, then i'd like you to update your comment in the code.  The primary thing it should be talking about is why this code is needed.  i.e. how we can get in a state when the syntax tree isn't in the containing compilation.  It should be clear that thsi is only for speculative cases, and that if you were to hit this in a non-speculative case that that would be a bug.  The comment about perf is interesting, but secondary.

Honestly... if we do go with this approach in the short term, i would augment 'ILocalSymbolInternal' to reveal if it was created from a SpeculativeSemanticModel.  And then i would assert in your code that if we have a syntax tree that doesn't belong to the compilation that the local must be a speculative symbol.

--

Finally: Is this only a concern for locals?  What about other interior symbols like "IRangeVariableSymbol".  Will this fix work for them as well?  Or is this only for locals?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi an alternate fix for this would be to have the symbol be able to indicate that it came for a semantic model and allow you to retrieve the speculative semantic model that created it, it similar to how ISourceAssemblySymbol allows you to get the compilation.

Oh... this sounds MUCH nicer.   Note: i think this could be done without any public changes since you're at the compiler layer here and you can access internal APIs.  Without any API changes it feels like this could be cheap enough to do now.  It doesn't sound particular hard either.  So is it something we can just pick up and do as part of this fix?

**Comment by jmarolf** [comment]
@CyrusNajmabadi When I first suggested this to @mattwar he got the heebie-jeebies.  I'll talk to him again and see if we can address his concerns about that plan.

**Comment by rchande** [comment]
@CyrusNajmabadi @jmarolf This is in the Workspaces project, does it have access to compiler internals? If so, I can totally try that approach.

@CyrusNajmabadi, yeah, this is actually done for locals, range variables, and labels.

**Comment by rchande** [comment]
No, it doesn't look like Workspaces can access compiler internals.

**Comment by rchande** [comment]
Cyrus, I added a better explanation of why this is happening.

**Comment by mattwar** [comment]
@jmarolf I don't think it is possible to have the symbol tell you what semantic model it came from.

**Comment by CyrusNajmabadi** [comment]
> @jmarolf I don't think it is possible to have the symbol tell you what semantic model it came from.

Why?  For local-like things, don't they always have to come from a semantic model?  SourceLocalSymbol points to a Compilation (through it's Binder instance).  Given that, and the tree it knows it comes from, are we sure we can't get back to the semantic model that created it?

**Comment by balajikris** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/7300/
// Retest reason: infra break

**Comment by balajikris** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/7338/
// Retest reason: infra break

**Comment by mattwar** [comment]
The local symbol has a reference to the binder that created it, which is likely the binder for the block that the symbol is declared within.  So I suppose you could find the ancestor that is the speculative binder (if that's really a separate type), and then create a speculative semantic model with that binder.  Note: the speculative binder has the original sub-tree node that is being speculated on.

**Comment by mattwar** [comment]
Also note: SourceLocalSymbol has a reference to the TypeSyntax of the declaration...  It's keeping the entire tree alive.  This is generally a bad thing to do.  It should have a SyntaxReference instead.

**Comment by rchande** [comment]
Did some more testing, turns out we won't show the tooltip for any occurrence of a local with multiple occurrences. The tooltip is blank at either instance of $$ below.

```C#
class C
{
    void foo()
    {
        System.Func<int> lambda = () =>
        {
            {
            int x;
                x$$
            }
            string x;
            x$$

            return 0;
        }
    }
}
```

**Comment by rchande** [comment]
@mattwar @CyrusNajmabadi @Pilchie: I'm not sure how to proceed from here. Do we want to take this to unblock dogfooding/update previews or work on a better fix?

**Comment by CyrusNajmabadi** [comment]
I'm ok with this going in, with a bug to track a better way of handling speculative symbols here.

**Comment by jmarolf** [comment]
:+1:

**Comment by rchande** [comment]
Filed https://github.com/dotnet/roslyn/issues/11436

**Pull request merged**